lib/Frontend/ParseableInterfaceSupport.cpp

```diff
@@ -104,24 +143,58 @@ ParseableInterfaceModuleLoader::configureSubInvocationAndOutputPath(
   FEOpts.InputsAndOutputs.setMainAndSupplementaryOutputs({MainOut}, {SOPs});
 }
 
-// FIXME: this needs to be a more extensive up-to-date check.
+// Write thee world's simplest dependencies file: a version identifier on
+// a line followed by a list of files, one per line.
+static bool writeSwiftInterfaceDeps(DiagnosticEngine &Diags,
+                                    ArrayRef<std::string> Deps,
+                                    StringRef DepPath) {
+  return withOutputFile(Diags, DepPath, [&](llvm::raw_pwrite_stream &out) {
+      out << SWIFT_INTERFACE_DEPS_VERSION << '\n';
+      for (auto const &D : Deps) {
+        out << D << '\n';
+      }
+      return false;
+    });
+}
+
+// Check that the output .swiftmodule file is at least as new as all the
+// dependencies it read when it was built last time.
 static bool
 swiftModuleIsUpToDate(clang::vfs::FileSystem &FS,
-                      StringRef InPath, StringRef OutPath) {
-  if (FS.exists(OutPath)) {
-    auto InStatus = FS.status(InPath);
+                      StringRef InPath, StringRef OutPath, StringRef DepPath) {
+  if (FS.exists(OutPath) && FS.exists(DepPath)) {
```

Comment:
Nit: Can you invert this check?

Reply:
Done (redid control flow to be a bunch of early false returns, ending in success).